### PR TITLE
fix: checks for cyclic dependencies

### LINKED ISSUE #2295: Cyclic dependencies cause nargo to stack overflow
### Aim

Nargo panics with a stack overflow when given the following TOML:
```toml
[package]
name = "baz"
type = "bin"
authors = [""]
compiler_version = "0.9.0"

[dependencies]
foo_bar = { path = "./" }
```

### Expected Behavior

Dependency paths should be checked for cycles and a proper error should be given instead of panicing with a stack overflow

### Bug

Panic instead of a good error message.

### To Reproduce

1.
2.
3.
4.

### Installation Method

None

### Nargo Version

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Maybe

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #2295 

## Summary\*

Keep track of the processed packages and check against this list before processing a new one.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/nargo_toml/src/errors.rs [hunk 1]
```diff
@@ -69,6 +69,9 @@ pub enum ManifestError {
 
     #[error(transparent)]
     SemverError(SemverError),
+
+    #[error("Cyclic package dependency found when processing {toml}")]
+    CyclicDependency { toml: PathBuf },
```
Comment (jfecher) [inline tooling/nargo_toml/src/errors.rs:74]: Can we attempt to show the user the cyclic dependency here? I think it'd help a lot while debugging.

**Comment by kevaundray** [comment]
@guipublic bumping this

File: tooling/nargo_toml/src/errors.rs (hunk 1 above)
Comment (guipublic) [reply tooling/nargo_toml/src/errors.rs]: The error displays the full cycle now.

**Comment by kevaundray** [comment]
Can we add a quick unit test for this?

**Comment by TomAFrench** [comment]
Could we add a rust test which shows that we throw a `CyclicDependency` error as well as the integration test? It's nice to have the integration test as check on behaviour but it doesn't show that the compilation is failing due to a cyclic dependency.

Should be fairly simple to set up a tempdir with two `Nargo.toml`s which point at each other and attempt to resolve them.

**Review by jfecher — APPROVED** [review]
LGTM.

Per @TomAFrench's comment. This could be a case where we consider golden tests https://github.com/noir-lang/noir/issues/3884 as well - at least in the long term. IMO it is nice having a unified way to check for both expected output and expected error output.

**Pull request merged**